src/Abp.Zero.EntityFramework/EntityHistory/EntitySnapshotManager.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.Entity;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+using Abp.Dependency;
+using Abp.Domain.Entities;
+using Abp.Domain.Repositories;
+using Abp.EntityFramework;
+using Abp.EntityFramework.Repositories;
+using Abp.Events.Bus.Entities;
+using Abp.Json;
+
+namespace Abp.EntityHistory
+{
+    public class EntitySnapshotManager : IEntitySnapshotManager, ITransientDependency
+    {
+        private readonly IRepository<EntityChange, long> _entityChangeRepository;
+
+        public EntitySnapshotManager(IRepository<EntityChange, long> entityChangeRepository)
+        {
+            _entityChangeRepository = entityChangeRepository;
+        }
+        public async Task<EntityHistorySnapshot> GetSnapshotAsync<TEntity>(int id, DateTime snapshotTime) where TEntity : class, IEntity<int>
+        {
+            return await GetSnapshotAsync<TEntity, int>(id, snapshotTime);
+        }
+
+        public async Task<EntityHistorySnapshot> GetSnapshotAsync<TEntity, TPrimaryKey>(TPrimaryKey id, DateTime snapshotTime) where TEntity : class, IEntity<TPrimaryKey>
+        {
+            var entity = _entityChangeRepository.GetDbContext()
+                .Set<TEntity>().AsQueryable().FirstOrDefault(CreateEqualityExpressionForId<TEntity, TPrimaryKey>(id));
+
+            var changedPropertiesDictionary = new Dictionary<string, string>();
+            var propertyChangesStackTreeDictionary = new Dictionary<string, string>();
+
+            if (entity != null)
+            {
+                string fullName = typeof(TEntity).FullName;
+                var idJson = id.ToJsonString();
+
+                var changes = await _entityChangeRepository.GetAll()//select all changes which created after snapshot time 
```

Comment:
do we need `Include(e => e.PropertChanges)` here?

Reply:
also, can extract this into 
```csharp
protected virtual IQueryable<EntityChange> GetEntityChanges<TEntity, TPrimaryKey>(TPrimaryKey id, DateTime snapshotTime)
{
    string fullName = typeof(TEntity).FullName;
    var idJson = id.ToJsonString();
    return _entityChangeRepository.GetAll()//select all changes which created after snapshot time 
                                  .Where(x => x.EntityTypeFullName == fullName && x.EntityId == idJson)
                                  .Where(x => x.ChangeTime > snapshotTime && x.ChangeType != EntityChangeType.Created);
}
```

Reply:
> do we need `Include(e => e.PropertChanges)` here?

No, i am selecting it in row 47